File: frontend/src/api/concepts.ts
import { client, handleError, withRetry } from '@/api/base'

export interface BundleListItem {
    bundle_index: number
    bundle_id?: string | null
    lifecycle?: string | null
    size: number
    venue_count: number
    venue_coverage: string[]
    top_entities: Array<{
        canonical_id?: string | null
        canonical_surface: string
        label: string
        node_key?: string | null
        paper_hits: number
    }>
    growth_indicator: 'growing' | 'declining' | 'stable'
}

export interface BundleDetail {
    bundle_index: number
    bundle_id?: string | null
    lifecycle?: string | null
    birth_year?: number | null
    latest_year?: number | null
    size: number
    latest_year_papers: number
    venue_count: number
    growth_rate: number
    cohesion: number
    internal_edge_weight: number
    external_edge_weight: number
    venue_coverage: string[]
    members: string[]
    top_entities: Array<{
        canonical_id?: string | null
        canonical_surface: string
        label: string
        node_key?: string | null
        paper_hits: number
    }>
    yearly_paper_counts: Record<string, number>
    previous_year_papers: number
}

export interface EntityNeighbor {
    canonical_id: string
    canonical_surface: string
    label: string
    paper_count: number
}

export interface CooccurrenceEdgeItem {
    left_node: string
    left_canonical_id?: string | null
    left_label: string
    right_node: string
    right_canonical_id?: string | null
    right_label: string
    paper_count: number
    venue: string
    year: number
}

export const conceptsApi = {
    async listBundles(limit = 100, offset = 0): Promise<BundleListItem[]> {
        try {
            const { data } = await withRetry(() =>
                client.get('/concepts/bundles', { params: { limit, offset } })
            )
            return data
        } catch (error) {
            return handleError(error)
        }
    },

    async getBundle(bundleIndex: number): Promise<BundleDetail> {
        try {
            const { data } = await withRetry(() => client.get(`/concepts/bundles/${bundleIndex}`))
            return data
        } catch (error) {
            return handleError(error)
        }
    },

    async getEntityNeighbors(canonicalId: string, params: {
        limit?: number; venue?: string; year?: number;
    } = {}): Promise<EntityNeighbor[]> {
        try {
            const { data } = await withRetry(() =>
                client.get(`/concepts/neighbors/${canonicalId}`, { params })
            )
            return data
        } catch (error) {
            return handleError(error)
        }
    },

    async getCooccurrenceEdges(params: {
        canonical_id?: string; venue?: string; year?: number;
        min_paper_count?: number; limit?: number; offset?: number;
    } = {}): Promise<CooccurrenceEdgeItem[]> {
        try {
            const { data } = await withRetry(() =>
                client.get('/concepts/edges', { params })
            )
            return data
        } catch (error) {
            return handleError(error)
        }
    },
}

File: backend/app/schemas/concepts.py
"""
Pydantic response schemas for concept graph endpoints.
"""

from typing import Literal

from pydantic import BaseModel


class BundleTopEntity(BaseModel):
    canonical_id: str | None = None
    canonical_surface: str
    label: str
    node_key: str | None = None
    paper_hits: int


class BundleListItem(BaseModel):
    """Summary of a concept bundle for the grid view."""

    bundle_index: int
    bundle_id: str | None = None
    lifecycle: str | None = None
    size: int
    venue_count: int
    venue_coverage: list[str]
    top_entities: list[BundleTopEntity]
    growth_indicator: Literal["growing", "declining", "stable"]

    model_config = {"from_attributes": True}


class BundleDetail(BaseModel):
    """Full detail for a single concept bundle."""

    bundle_index: int
    bundle_id: str | None = None
    lifecycle: str | None = None
    birth_year: int | None = None
    latest_year: int | None = None
    size: int
    latest_year_papers: int
    venue_count: int
    growth_rate: float
    cohesion: float
    internal_edge_weight: int
    external_edge_weight: int
    venue_coverage: list[str]
    members: list[str]
    top_entities: list[BundleTopEntity]
    yearly_paper_counts: dict[str, int]
    previous_year_papers: int

    model_config = {"from_attributes": True}


class CooccurrenceEdgeItem(BaseModel):
    """A co-occurrence edge between two entities."""

    left_node: str
    left_canonical_id: str | None = None
    left_label: str
    right_node: str
    right_canonical_id: str | None = None
    right_label: str
    paper_count: int
    venue: str
    year: int

    model_config = {"from_attributes": True}


class EntityNeighbor(BaseModel):
    """A neighbor entity in the co-occurrence graph."""

    canonical_id: str
    canonical_surface: str
    label: str
    paper_count: int  # co-occurrence count
